File: js/src/store/json-store.ts
/**
 * hanimo-rag v2 — JSON File Store
 */

import * as fs from 'node:fs'
import * as path from 'node:path'
import type { IndexedChunk, StoreStats } from '../types.js'
import { StoreBase } from './base.js'

interface StoreData {
  version: number
  chunks: IndexedChunk[]
}

export class JsonStore extends StoreBase {
  private filePath: string
  private data: StoreData

  constructor(storePath: string) {
    super()
    // Ensure directory exists
    if (!fs.existsSync(storePath)) {
      fs.mkdirSync(storePath, { recursive: true })
    }
    this.filePath = path.join(storePath, 'index.json')
    this.data = this.load()
  }

  private load(): StoreData {
    if (!fs.existsSync(this.filePath)) {
      return { version: 2, chunks: [] }
    }
    try {
      const raw = fs.readFileSync(this.filePath, 'utf-8')
      const parsed = JSON.parse(raw) as StoreData
      if (!parsed.chunks || !Array.isArray(parsed.chunks)) {
        return { version: 2, chunks: [] }
      }
      return parsed
    } catch {
      return { version: 2, chunks: [] }
    }
  }

  private save(): void {
    fs.writeFileSync(this.filePath, JSON.stringify(this.data, null, 2), 'utf-8')
  }

  saveChunks(chunks: IndexedChunk[]): number {
    // Deduplicate by id — overwrite existing chunks with same id
    const existingIds = new Set(this.data.chunks.map((c) => c.id))
    const newChunks: IndexedChunk[] = []

    for (const chunk of chunks) {
      if (existingIds.has(chunk.id)) {
        // Replace existing
        const idx = this.data.chunks.findIndex((c) => c.id === chunk.id)
        if (idx !== -1) {
          this.data.chunks[idx] = chunk
        }
      } else {
        newChunks.push(chunk)
      }
    }

    this.data.chunks.push(...newChunks)
    this.save()
    return chunks.length
  }

  lookup(keys: string[], category?: string, limit: number = 10): IndexedChunk[] {
    if (keys.length === 0) {
      return []
    }

    const normalizedKeys = keys.map((k) => k.toLowerCase().trim())

    const scored: Array<{ chunk: IndexedChunk; score: number; matchedKeys: string[] }> = []

    for (const chunk of this.data.chunks) {
      // Filter by category if specified
      if (category && chunk.category.toLowerCase() !== category.toLowerCase()) {
        continue
      }

      const allChunkKeys = [
        ...chunk.topics,
        ...chunk.entities,
        ...chunk.questions,
      ].map((k) => k.toLowerCase().trim())

      let score = 0
      const matchedKeys: string[] = []

      for (const searchKey of normalizedKeys) {
        for (const chunkKey of allChunkKeys) {
          if (chunkKey.includes(searchKey) || searchKey.includes(chunkKey)) {
            score++
            matchedKeys.push(chunkKey)
            break // Count each search key at most once per chunk
          }
        }
      }

      // Also check content for partial matches (lower weight)
      const contentLower = chunk.content.toLowerCase()
      for (const searchKey of normalizedKeys) {
        if (contentLower.includes(searchKey)) {
          score += 0.5
        }
      }

      if (score > 0) {
        scored.push({ chunk, score, matchedKeys: [...new Set(matchedKeys)] })
      }
    }

    // Sort by score descending
    scored.sort((a, b) => b.score - a.score)

    return scored.slice(0, limit).map((s) => s.chunk)
  }

  getAllKeys(): Set<string> {
    const keys = new Set<string>()
    for (const chunk of this.data.chunks) {
      for (const t of chunk.topics) keys.add(t.toLowerCase())
      for (const e of chunk.entities) keys.add(e.toLowerCase())
      for (const q of chunk.questions) keys.add(q.toLowerCase())
    }
    return keys
  }

  getStats(): StoreStats {
    const sources = new Set(this.data.chunks.map((c) => c.source))

    // Count topic frequency
    const topicCounts = new Map<string, number>()
    for (const chunk of this.data.chunks) {
      for (const topic of chunk.topics) {
        const key = topic.toLowerCase()
        topicCounts.set(key, (topicCounts.get(key) ?? 0) + 1)
      }
    }

    const topTopics = [...topicCounts.entries()]
      .sort((a, b) => b[1] - a[1])
      .slice(0, 10)
      .map(([topic]) => topic)

    return {
      totalChunks: this.data.chunks.length,
      totalSources: sources.size,
      topTopics,
    }
  }

  deleteBySource(source: string): number {
    const before = this.data.chunks.length
    this.data.chunks = this.data.chunks.filter((c) => c.source !== source)
    const deleted = before - this.data.chunks.length
    if (deleted > 0) {
      this.save()
    }
    return deleted
  }
}

File: hanimo_rag/store/json_store.py
"""JSON file-based storage backend."""
from __future__ import annotations

import json
import logging
from pathlib import Path

from hanimo_rag.store.base import StoreBase
from hanimo_rag.types import IndexedChunk

logger = logging.getLogger(__name__)


class JsonStore(StoreBase):
    """Simple JSON file store. Good for <100k chunks."""

    def __init__(self, store_path: str) -> None:
        self.store_dir = Path(store_path)
        self.store_dir.mkdir(parents=True, exist_ok=True)
        self.index_file = self.store_dir / "index.json"
        self._chunks: list[IndexedChunk] = []
        self._load()

    def _load(self) -> None:
        if self.index_file.exists():
            try:
                data = json.loads(self.index_file.read_text(encoding="utf-8"))
                self._chunks = [IndexedChunk.from_dict(d) for d in data]
                logger.info("Loaded %d chunks from %s", len(self._chunks), self.index_file)
            except (json.JSONDecodeError, KeyError) as exc:
                logger.warning("Failed to load index: %s — starting fresh", exc)
                self._chunks = []
        else:
            self._chunks = []

    def _save(self) -> None:
        data = [c.to_dict() for c in self._chunks]
        self.index_file.write_text(
            json.dumps(data, ensure_ascii=False, indent=2),
            encoding="utf-8",
        )

    def save_chunks(self, chunks: list[IndexedChunk]) -> int:
        existing_ids = {c.id for c in self._chunks}
        added = 0
        for chunk in chunks:
            if chunk.id not in existing_ids:
                self._chunks.append(chunk)
                existing_ids.add(chunk.id)
                added += 1
        self._save()
        return added

    def lookup(
        self,
        keys: list[str],
        category: str | None = None,
        limit: int = 20,
    ) -> list[IndexedChunk]:
        if not keys:
            return []

        lower_keys = [k.lower() for k in keys]
        scored: list[tuple[float, IndexedChunk]] = []

        for chunk in self._chunks:
            if category and chunk.category != category:
                continue

            score = _score_chunk(chunk, lower_keys)
            if score > 0:
                scored.append((score, chunk))

        scored.sort(key=lambda x: x[0], reverse=True)
        return [chunk for _, chunk in scored[:limit]]

    def get_all_keys(self) -> set[str]:
        keys: set[str] = set()
        for chunk in self._chunks:
            keys.update(chunk.topics)
            keys.update(chunk.entities)
        return keys

    def get_stats(self) -> dict:
        sources = {c.source for c in self._chunks}
        categories: dict[str, int] = {}
        for c in self._chunks:
            categories[c.category] = categories.get(c.category, 0) + 1
        return {
            "total_chunks": len(self._chunks),
            "sources": len(sources),
            "categories": categories,
            "store_type": "json",
            "store_path": str(self.store_dir),
        }

    def delete_by_source(self, source: str) -> int:
        before = len(self._chunks)
        self._chunks = [c for c in self._chunks if c.source != source]
        deleted = before - len(self._chunks)
        if deleted > 0:
            self._save()
        return deleted


def _score_chunk(chunk: IndexedChunk, lower_keys: list[str]) -> float:
    """Score a chunk by keyword overlap with topics, entities, and questions."""
    score = 0.0

    chunk_topics = [t.lower() for t in chunk.topics]
    chunk_entities = [e.lower() for e in chunk.entities]
    chunk_questions = " ".join(chunk.questions).lower()

    for key in lower_keys:
        # Exact match in topics (highest weight)
        if key in chunk_topics:
            score += 3.0
        # Partial match in topics
        elif any(key in t or t in key for t in chunk_topics):
            score += 1.5

        # Exact match in entities
        if key in chunk_entities:
            score += 2.5
        # Partial match in entities
        elif any(key in e or e in key for e in chunk_entities):
            score += 1.0

        # Match in questions
        if key in chunk_questions:
            score += 1.0

        # Match in summary (lowest weight)
        if key in chunk.summary.lower():
            score += 0.5

    return score
